{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "import os, glob\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import Bio\n",
    "from Bio.Seq import MutableSeq, Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import GC\n",
    "import fnmatch\n",
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'primer' from '/home/dave/Documents/Bacteria/libs/primer.py'>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.chdir('/home/dave/Documents/Bacteria/libs')\n",
    "import primer #From DaveBio\n",
    "import Bacterial #From DaveBio\n",
    "import PyClustalW #From DaveBio\n",
    "import probe #From DaveBio\n",
    "primer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define Common Primers\n",
    "\n",
    "#Forward Primers \n",
    "F8 = 'AGAGTTTGATCCTGGCTCAG'\n",
    "F27 = 'AGAGTTTGATCMTGGCTCAG'\n",
    "CC_F = 'CCAGACTCCTACGGGAGGCAGC'\n",
    "F357F = 'CTCCTACGGGAGGCAGCAG'\n",
    "F515F = 'GTGCCAGCMGCCGCGGTAA'\n",
    "F533F = 'GTGCCAGCAGCCGCGGTAA'\n",
    "F16S = 'CAACGAGCGCAACCCT'\n",
    "F1237F = 'GGGCTACACACGYGCWAC'\n",
    "V1F = 'GYGGCGNACGGGTGAGTAA'\n",
    "\n",
    "#Reverse Primers\n",
    "R519 = 'GWATTACCGCGGCKGCTG'\n",
    "CDR = 'CTTGTGCGGGCCCCCGTCAATTC'\n",
    "R907 = 'CCGTCAATTCMTTTRAGTTT'\n",
    "R1100 = 'AGGGTTGCGCTCGTTG'\n",
    "R1391 = 'GACGGGCGGTGTGTRCA'\n",
    "R1492 = 'GGTTACCTTGTTACGACTT'\n",
    "R1492 = 'ACCTTGTTACGACTT'\n",
    "V6R = 'AGCTGACGACANCCATGCA'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Record id</th>\n",
       "      <th>Forward Primer</th>\n",
       "      <th>forward_primer_position</th>\n",
       "      <th>Reverse Primer</th>\n",
       "      <th>reverse_primer_position</th>\n",
       "      <th>GC Content</th>\n",
       "      <th>Length of Amplicon</th>\n",
       "      <th>Amplicon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Micrococcus.luteus</td>\n",
       "      <td>NR_037113.1</td>\n",
       "      <td>GTGGCGAACGGGTGAGTAA</td>\n",
       "      <td>75</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>1009</td>\n",
       "      <td>57.3</td>\n",
       "      <td>953</td>\n",
       "      <td>GTGGCGAACGGGTGAGTAACACGTGAGTAACCTGCCCTTAACTCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Escherichia.coli</td>\n",
       "      <td>AJ605115.1</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAA</td>\n",
       "      <td>30</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>981</td>\n",
       "      <td>55.4</td>\n",
       "      <td>970</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCTGATGGAGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Streptococcus.pyogenes</td>\n",
       "      <td>NR_028598.1</td>\n",
       "      <td>N/a</td>\n",
       "      <td>N/a</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>1013</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Streptococcus.pneumoniae</td>\n",
       "      <td>NR_028665.1</td>\n",
       "      <td>N/a</td>\n",
       "      <td>N/a</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>1046</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Streptococcus.agalactiae</td>\n",
       "      <td>NR_040821.1</td>\n",
       "      <td>N/a</td>\n",
       "      <td>N/a</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>1031</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Escherichia.coli</td>\n",
       "      <td>XX000001.1</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAA</td>\n",
       "      <td>100</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>1050</td>\n",
       "      <td>55.4</td>\n",
       "      <td>969</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCTGATGGAGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Haemophilus.influenzae</td>\n",
       "      <td>NR_044682.2</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAA</td>\n",
       "      <td>100</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>1048</td>\n",
       "      <td>52.5</td>\n",
       "      <td>967</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAATGCTTGGGAATCTGGCTTATGGAGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Moraxella.osloensis</td>\n",
       "      <td>NR_104936.1</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAA</td>\n",
       "      <td>91</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>1036</td>\n",
       "      <td>51.1</td>\n",
       "      <td>964</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAACATTTAGGAATCTACCTAGTAGTGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Serratia.marcescens</td>\n",
       "      <td>NR_036886.1</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAA</td>\n",
       "      <td>95</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>1045</td>\n",
       "      <td>54.9</td>\n",
       "      <td>969</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCTGATGGAGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Flavonifractor.plautii</td>\n",
       "      <td>NR_029356.1</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAA</td>\n",
       "      <td>92</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>1027</td>\n",
       "      <td>55.0</td>\n",
       "      <td>954</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAACGCGTGAGGAACCTGCCTTGGAGAGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Agrobacterium.tumefaciens</td>\n",
       "      <td>NR_041396.1</td>\n",
       "      <td>N/a</td>\n",
       "      <td>N/a</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>962</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Escherichia.coli</td>\n",
       "      <td>NR_024570.1</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAA</td>\n",
       "      <td>92</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>1041</td>\n",
       "      <td>54.9</td>\n",
       "      <td>968</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCTGATGGAGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Staphylococcus.epidermidis</td>\n",
       "      <td>JX131632.1</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAA</td>\n",
       "      <td>68</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>1031</td>\n",
       "      <td>51.5</td>\n",
       "      <td>982</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAACACGTGGATAACCTACCTATAAGACTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Corynebacterium.amycolatum</td>\n",
       "      <td>NR_026215.1</td>\n",
       "      <td>GTGGCGAACGGGTGAGTAA</td>\n",
       "      <td>69</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>1000</td>\n",
       "      <td>57.6</td>\n",
       "      <td>950</td>\n",
       "      <td>GTGGCGAACGGGTGAGTAACACGTGGGTGACCTGCCCTGCACTTCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Listeria.monocytogenes</td>\n",
       "      <td>NR_044823.1</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAA</td>\n",
       "      <td>97</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>1056</td>\n",
       "      <td>54.0</td>\n",
       "      <td>978</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAACACGTGGGCAACCTGCCTGTAAGTTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Citrobacter.freundii</td>\n",
       "      <td>NR_028894.1</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAA</td>\n",
       "      <td>81</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>1031</td>\n",
       "      <td>55.9</td>\n",
       "      <td>969</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCCGATGGAGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Enterococcus.faecalis</td>\n",
       "      <td>NR_040789.1</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAA</td>\n",
       "      <td>102</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>1059</td>\n",
       "      <td>53.8</td>\n",
       "      <td>976</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAACACGTGGGTAACCTACCCATCAGAGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Staphylococcus.pettenkoferi</td>\n",
       "      <td>NR_114596.1</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAA</td>\n",
       "      <td>92</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>1054</td>\n",
       "      <td>53.0</td>\n",
       "      <td>981</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAACACGTGGGTAACCTACCTATAAGACTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Klebsiella.pneumoniae</td>\n",
       "      <td>NR_036794.1</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAA</td>\n",
       "      <td>92</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>1043</td>\n",
       "      <td>55.8</td>\n",
       "      <td>970</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCTGATGGAGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Staphylococcus.aureus</td>\n",
       "      <td>NR_118997.2</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAA</td>\n",
       "      <td>99</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>1061</td>\n",
       "      <td>51.3</td>\n",
       "      <td>981</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAACACGTGGATAACCTACCTATAAGACTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Staphylococcus.carnosus</td>\n",
       "      <td>LC145699.1</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAA</td>\n",
       "      <td>36</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>998</td>\n",
       "      <td>52.9</td>\n",
       "      <td>981</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAACACGTGGGTAACCTACCTATAAGACTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Neisseria.gonorrhoeae</td>\n",
       "      <td>NR_026079.2</td>\n",
       "      <td>GTGGCGAACGGGTGAGTAA</td>\n",
       "      <td>101</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>1051</td>\n",
       "      <td>54.9</td>\n",
       "      <td>969</td>\n",
       "      <td>GTGGCGAACGGGTGAGTAACATATCGGAACGTACCGGGTAGCGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Enterobacter.cloacae</td>\n",
       "      <td>NR_102794.2</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAA</td>\n",
       "      <td>99</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>1052</td>\n",
       "      <td>55.6</td>\n",
       "      <td>972</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCTGATGGAGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Klebsiella.oxytoca</td>\n",
       "      <td>NR_041749.1</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAA</td>\n",
       "      <td>84</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>1034</td>\n",
       "      <td>55.3</td>\n",
       "      <td>969</td>\n",
       "      <td>GTGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCCGATGGAGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Neisseria.meningitidis</td>\n",
       "      <td>NR_104946.1</td>\n",
       "      <td>GTGGCGAACGGGTGAGTAA</td>\n",
       "      <td>37</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>987</td>\n",
       "      <td>54.1</td>\n",
       "      <td>969</td>\n",
       "      <td>GTGGCGAACGGGTGAGTAACATATCGGAACGTACCGAGTAGTGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Streptococcus.mitis</td>\n",
       "      <td>NR_028664.1</td>\n",
       "      <td>N/a</td>\n",
       "      <td>N/a</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>1052</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Pseudomonas.aeruginosa</td>\n",
       "      <td>NR_026078.1</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAA</td>\n",
       "      <td>98</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>1046</td>\n",
       "      <td>53.6</td>\n",
       "      <td>967</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAATGCCTAGGAATCTGCCTGGTAGTGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Streptococcus.dysgalactiae</td>\n",
       "      <td>NR_027517.1</td>\n",
       "      <td>N/a</td>\n",
       "      <td>N/a</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>1051</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Streptococcus.parasanguinis</td>\n",
       "      <td>NR_024842.1</td>\n",
       "      <td>N/a</td>\n",
       "      <td>N/a</td>\n",
       "      <td>TGCATGGTTGTCGTCAGCT</td>\n",
       "      <td>1046</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Fusobacterium.necrophorum</td>\n",
       "      <td>NR_042365.1</td>\n",
       "      <td>GTGGCGCACGGGTGAGTAA</td>\n",
       "      <td>73</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>1007</td>\n",
       "      <td>52.0</td>\n",
       "      <td>953</td>\n",
       "      <td>GTGGCGCACGGGTGAGTAACGCGTAAAGAACTTGCCTCTTAGACTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Chlamydia.trachomatis</td>\n",
       "      <td>NR_025888.1</td>\n",
       "      <td>N/a</td>\n",
       "      <td>N/a</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>1056</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Acinetobacter.baumannii</td>\n",
       "      <td>NR_026206.1</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAA</td>\n",
       "      <td>64</td>\n",
       "      <td>TGCATGGCTGTCGTCAGCT</td>\n",
       "      <td>1015</td>\n",
       "      <td>52.4</td>\n",
       "      <td>970</td>\n",
       "      <td>GCGGCGGACGGGTGAGTAATGCTTAGGAATCTGCCTATTAGTGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Bifidobacterium.adolescentis</td>\n",
       "      <td>NR_074802.2</td>\n",
       "      <td>N/a</td>\n",
       "      <td>N/a</td>\n",
       "      <td>N/a</td>\n",
       "      <td>N/a</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Species    Record id       Forward Primer  \\\n",
       "0             Micrococcus.luteus  NR_037113.1  GTGGCGAACGGGTGAGTAA   \n",
       "1               Escherichia.coli   AJ605115.1  GTGGCGGACGGGTGAGTAA   \n",
       "2         Streptococcus.pyogenes  NR_028598.1                  N/a   \n",
       "3       Streptococcus.pneumoniae  NR_028665.1                  N/a   \n",
       "4       Streptococcus.agalactiae  NR_040821.1                  N/a   \n",
       "5               Escherichia.coli   XX000001.1  GTGGCGGACGGGTGAGTAA   \n",
       "6         Haemophilus.influenzae  NR_044682.2  GTGGCGGACGGGTGAGTAA   \n",
       "7            Moraxella.osloensis  NR_104936.1  GTGGCGGACGGGTGAGTAA   \n",
       "8            Serratia.marcescens  NR_036886.1  GCGGCGGACGGGTGAGTAA   \n",
       "9         Flavonifractor.plautii  NR_029356.1  GTGGCGGACGGGTGAGTAA   \n",
       "10     Agrobacterium.tumefaciens  NR_041396.1                  N/a   \n",
       "11              Escherichia.coli  NR_024570.1  GTGGCGGACGGGTGAGTAA   \n",
       "12    Staphylococcus.epidermidis   JX131632.1  GCGGCGGACGGGTGAGTAA   \n",
       "13    Corynebacterium.amycolatum  NR_026215.1  GTGGCGAACGGGTGAGTAA   \n",
       "14        Listeria.monocytogenes  NR_044823.1  GTGGCGGACGGGTGAGTAA   \n",
       "15          Citrobacter.freundii  NR_028894.1  GTGGCGGACGGGTGAGTAA   \n",
       "16         Enterococcus.faecalis  NR_040789.1  GTGGCGGACGGGTGAGTAA   \n",
       "17   Staphylococcus.pettenkoferi  NR_114596.1  GCGGCGGACGGGTGAGTAA   \n",
       "18         Klebsiella.pneumoniae  NR_036794.1  GCGGCGGACGGGTGAGTAA   \n",
       "19         Staphylococcus.aureus  NR_118997.2  GCGGCGGACGGGTGAGTAA   \n",
       "20       Staphylococcus.carnosus   LC145699.1  GCGGCGGACGGGTGAGTAA   \n",
       "21         Neisseria.gonorrhoeae  NR_026079.2  GTGGCGAACGGGTGAGTAA   \n",
       "22          Enterobacter.cloacae  NR_102794.2  GTGGCGGACGGGTGAGTAA   \n",
       "23            Klebsiella.oxytoca  NR_041749.1  GTGGCGGACGGGTGAGTAA   \n",
       "24        Neisseria.meningitidis  NR_104946.1  GTGGCGAACGGGTGAGTAA   \n",
       "25           Streptococcus.mitis  NR_028664.1                  N/a   \n",
       "26        Pseudomonas.aeruginosa  NR_026078.1  GCGGCGGACGGGTGAGTAA   \n",
       "27    Streptococcus.dysgalactiae  NR_027517.1                  N/a   \n",
       "28   Streptococcus.parasanguinis  NR_024842.1                  N/a   \n",
       "29     Fusobacterium.necrophorum  NR_042365.1  GTGGCGCACGGGTGAGTAA   \n",
       "30         Chlamydia.trachomatis  NR_025888.1                  N/a   \n",
       "31       Acinetobacter.baumannii  NR_026206.1  GCGGCGGACGGGTGAGTAA   \n",
       "32  Bifidobacterium.adolescentis  NR_074802.2                  N/a   \n",
       "\n",
       "   forward_primer_position       Reverse Primer reverse_primer_position  \\\n",
       "0                       75  TGCATGGTTGTCGTCAGCT                    1009   \n",
       "1                       30  TGCATGGCTGTCGTCAGCT                     981   \n",
       "2                      N/a  TGCATGGTTGTCGTCAGCT                    1013   \n",
       "3                      N/a  TGCATGGTTGTCGTCAGCT                    1046   \n",
       "4                      N/a  TGCATGGTTGTCGTCAGCT                    1031   \n",
       "5                      100  TGCATGGCTGTCGTCAGCT                    1050   \n",
       "6                      100  TGCATGGCTGTCGTCAGCT                    1048   \n",
       "7                       91  TGCATGGCTGTCGTCAGCT                    1036   \n",
       "8                       95  TGCATGGCTGTCGTCAGCT                    1045   \n",
       "9                       92  TGCATGGTTGTCGTCAGCT                    1027   \n",
       "10                     N/a  TGCATGGCTGTCGTCAGCT                     962   \n",
       "11                      92  TGCATGGCTGTCGTCAGCT                    1041   \n",
       "12                      68  TGCATGGTTGTCGTCAGCT                    1031   \n",
       "13                      69  TGCATGGTTGTCGTCAGCT                    1000   \n",
       "14                      97  TGCATGGTTGTCGTCAGCT                    1056   \n",
       "15                      81  TGCATGGCTGTCGTCAGCT                    1031   \n",
       "16                     102  TGCATGGTTGTCGTCAGCT                    1059   \n",
       "17                      92  TGCATGGTTGTCGTCAGCT                    1054   \n",
       "18                      92  TGCATGGCTGTCGTCAGCT                    1043   \n",
       "19                      99  TGCATGGTTGTCGTCAGCT                    1061   \n",
       "20                      36  TGCATGGTTGTCGTCAGCT                     998   \n",
       "21                     101  TGCATGGCTGTCGTCAGCT                    1051   \n",
       "22                      99  TGCATGGCTGTCGTCAGCT                    1052   \n",
       "23                      84  TGCATGGCTGTCGTCAGCT                    1034   \n",
       "24                      37  TGCATGGCTGTCGTCAGCT                     987   \n",
       "25                     N/a  TGCATGGTTGTCGTCAGCT                    1052   \n",
       "26                      98  TGCATGGCTGTCGTCAGCT                    1046   \n",
       "27                     N/a  TGCATGGTTGTCGTCAGCT                    1051   \n",
       "28                     N/a  TGCATGGTTGTCGTCAGCT                    1046   \n",
       "29                      73  TGCATGGCTGTCGTCAGCT                    1007   \n",
       "30                     N/a  TGCATGGCTGTCGTCAGCT                    1056   \n",
       "31                      64  TGCATGGCTGTCGTCAGCT                    1015   \n",
       "32                     N/a                  N/a                     N/a   \n",
       "\n",
       "    GC Content  Length of Amplicon  \\\n",
       "0         57.3                 953   \n",
       "1         55.4                 970   \n",
       "2          0.0                   0   \n",
       "3          0.0                   0   \n",
       "4          0.0                   0   \n",
       "5         55.4                 969   \n",
       "6         52.5                 967   \n",
       "7         51.1                 964   \n",
       "8         54.9                 969   \n",
       "9         55.0                 954   \n",
       "10         0.0                   0   \n",
       "11        54.9                 968   \n",
       "12        51.5                 982   \n",
       "13        57.6                 950   \n",
       "14        54.0                 978   \n",
       "15        55.9                 969   \n",
       "16        53.8                 976   \n",
       "17        53.0                 981   \n",
       "18        55.8                 970   \n",
       "19        51.3                 981   \n",
       "20        52.9                 981   \n",
       "21        54.9                 969   \n",
       "22        55.6                 972   \n",
       "23        55.3                 969   \n",
       "24        54.1                 969   \n",
       "25         0.0                   0   \n",
       "26        53.6                 967   \n",
       "27         0.0                   0   \n",
       "28         0.0                   0   \n",
       "29        52.0                 953   \n",
       "30         0.0                   0   \n",
       "31        52.4                 970   \n",
       "32         0.0                   0   \n",
       "\n",
       "                                             Amplicon  \n",
       "0   GTGGCGAACGGGTGAGTAACACGTGAGTAACCTGCCCTTAACTCTG...  \n",
       "1   GTGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCTGATGGAGGGG...  \n",
       "2                                                      \n",
       "3                                                      \n",
       "4                                                      \n",
       "5   GTGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCTGATGGAGGGG...  \n",
       "6   GTGGCGGACGGGTGAGTAATGCTTGGGAATCTGGCTTATGGAGGGG...  \n",
       "7   GTGGCGGACGGGTGAGTAACATTTAGGAATCTACCTAGTAGTGGGG...  \n",
       "8   GCGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCTGATGGAGGGG...  \n",
       "9   GTGGCGGACGGGTGAGTAACGCGTGAGGAACCTGCCTTGGAGAGGG...  \n",
       "10                                                     \n",
       "11  GTGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCTGATGGAGGGG...  \n",
       "12  GCGGCGGACGGGTGAGTAACACGTGGATAACCTACCTATAAGACTG...  \n",
       "13  GTGGCGAACGGGTGAGTAACACGTGGGTGACCTGCCCTGCACTTCG...  \n",
       "14  GTGGCGGACGGGTGAGTAACACGTGGGCAACCTGCCTGTAAGTTGG...  \n",
       "15  GTGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCCGATGGAGGGG...  \n",
       "16  GTGGCGGACGGGTGAGTAACACGTGGGTAACCTACCCATCAGAGGG...  \n",
       "17  GCGGCGGACGGGTGAGTAACACGTGGGTAACCTACCTATAAGACTG...  \n",
       "18  GCGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCTGATGGAGGGG...  \n",
       "19  GCGGCGGACGGGTGAGTAACACGTGGATAACCTACCTATAAGACTG...  \n",
       "20  GCGGCGGACGGGTGAGTAACACGTGGGTAACCTACCTATAAGACTG...  \n",
       "21  GTGGCGAACGGGTGAGTAACATATCGGAACGTACCGGGTAGCGGGG...  \n",
       "22  GTGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCTGATGGAGGGG...  \n",
       "23  GTGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCCGATGGAGGGG...  \n",
       "24  GTGGCGAACGGGTGAGTAACATATCGGAACGTACCGAGTAGTGGGG...  \n",
       "25                                                     \n",
       "26  GCGGCGGACGGGTGAGTAATGCCTAGGAATCTGCCTGGTAGTGGGG...  \n",
       "27                                                     \n",
       "28                                                     \n",
       "29  GTGGCGCACGGGTGAGTAACGCGTAAAGAACTTGCCTCTTAGACTG...  \n",
       "30                                                     \n",
       "31  GCGGCGGACGGGTGAGTAATGCTTAGGAATCTGCCTATTAGTGGGG...  \n",
       "32                                                     "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Generate Amplicons of Bacterial Species from Primers\n",
    "core_data = Bacterial.get_species()\n",
    "\n",
    "forward_primer = MutableSeq(V1F)\n",
    "reverse_primer = MutableSeq(V6R)\n",
    "\n",
    "fwd_tup = primer.degenerate_primer(forward_primer)\n",
    "rev_tup = primer.degenerate_primer(reverse_primer)\n",
    "calculated_data = primer.create_PCR_amplicon(core_data, rev_tup, fwd_tup)\n",
    "#calculated_data.to_csv('Bacterial_Amplicons.csv', sep='\\t')\n",
    "calculated_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                            ____________ NR_037113.1\n",
      "                                     ______|\n",
      "                           _________|      |_____________ NR_026215.1\n",
      "                          |         |\n",
      "                          |         |____________________ NR_074802.2\n",
      "                          |\n",
      "                          |                       ______ NR_028598.1\n",
      "                          |                     ,|\n",
      "                          |                    _||______ NR_040821.1\n",
      "                          |                   | |\n",
      "                          |                   | |_____ NR_027517.1\n",
      "                          |         __________|\n",
      "                          |        |          |      , NR_028665.1\n",
      "                         ,|        |          |   ___|\n",
      "                         ||        |          |__|   | NR_028664.1\n",
      "                         ||        |             |\n",
      "                         ||        |             |____ NR_024842.1\n",
      "                         ||  ______|\n",
      "                         || |      |               __ JX131632.1\n",
      "                         || |      |             _|\n",
      "                         || |      |            | |________ LC145699.1\n",
      "                         || |      |    ________|\n",
      "                         || |      |   |        |  __ NR_114596.1\n",
      "                        _|| |      |   |        |_|\n",
      "                       | ||_|      |___|          |___ NR_118997.2\n",
      "                       | |  |          |\n",
      "                       | |  |          |   ____________ NR_044823.1\n",
      "                       | |  |          |__|\n",
      "                       | |  |             |___________ NR_040789.1\n",
      "                     __| |  |\n",
      "                    |  | |  |  ________________________ NR_029356.1\n",
      "                    |  | |  |_|\n",
      "                    |  | |    |_____________________________ NR_042365.1\n",
      "                    |  | |\n",
      "                  __|  | |_____________________________________ NR_025888.1\n",
      "                 |  |  |\n",
      "                 |  |  |______________________________ NR_041396.1\n",
      "                 |  |\n",
      "                 |  |                  ____ NR_026079.2\n",
      "           ______|  |_________________|\n",
      "          |      |                    |__________ NR_104946.1\n",
      "          |      |\n",
      "          |      |      _____________ NR_104936.1\n",
      "   _______|      | ____|\n",
      "  |       |      ||    |______________ NR_026206.1\n",
      "  |       |       |\n",
      "  |       |       |_________________ NR_026078.1\n",
      " ,|       |\n",
      " ||       |___________________ NR_044682.2\n",
      " ||\n",
      " ||     __________ AJ605115.1\n",
      " ||   ,|\n",
      " ||___||_ NR_024570.1\n",
      " |    |\n",
      " |    |_ XX000001.1\n",
      " |\n",
      " |   _ NR_028894.1\n",
      "_|__|\n",
      " |  |_ NR_041749.1\n",
      " |\n",
      " |___ NR_036886.1\n",
      " |\n",
      " |__ NR_036794.1\n",
      " |\n",
      " |___ NR_102794.2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Generate Dendro of Bacterial Species from 16S Segments\n",
    "file_name = \"16S\"\n",
    "writen_alignment = PyClustalW.create_alignment_file(core_data, '16S Sequence', file_name)\n",
    "writen_alignment\n",
    "PyClustalW.ClustalW_alignment(str(file_name + '.faa'))\n",
    "#ClustalW_alignment\n",
    "PyClustalW.Phylo_tree(file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'----TGAGAATTTGATCTTGGTTCAGATTGAACGCTGGCGGCGTG----GATGAGGCATGC--AAGTCGAACGGA--------------GCAATTGTTTCGACG-----ATTGTTTAGTGGCGGAAGGGTTAGTAATGCATAGATAATTTGTCCTTAACTTGGGAATAACGGTTGGAAACGGCCGCTAATACCGAATGTGGCGATATTTGGGCATCCGAGTAACGTTAAAGAAGGGGATCTTAGGACCTTTCGGTTAAGGGAGAGTCTATGTGATATCAGCTAGTTGGTGGGGTAAAGGCCTACCAAGGCTATGACGTCTAGGCGGATTGAGAGATTGGCCGCCAACACTGGGACTGAGACACTGCCCAGACTCCTACGGGAGGCTGCAGTCGAGAATCTTTCGCAATGGAC-GGAAGTCTGACGAAGCGACGCCGCGTGTGTGATGAAGGCTCTAGGGTTGTAAAGCACTTTCGCTTGGGAATAAGAGAAGACGGTTAATACCCGCTGG-ATTTGAGCGTACCA-GGTAAAGAAGCACCGGCTAACTCCGTGCCAGCAGCTGCGGTAAT-ACGGAGGGTGCTAGCGTTAATCGGATTTATTGGGCGTAAAGGGCGTGTAGGCGGAAAGGTAAGTTAGTTGTCAAAGATCGGGGCTCAACCCCGAGTCGGCATCTAATACTATTT-TTCTAGAGGATAGATGGAGAAAAGGGAATTTCACGTG-TAGCGGTGAAATGCGTAGATATGTGGAAGAACACCAGTGGCGAAGGCGCTTTTCTAATTTATACCTGACGCTAA-GGCGCGAAAGCAAGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCTTGCCGTAAACGATGCATACTT--GATGTGGATGGTCTCAACCCC-ATCCGTGTCGGAG-CTAACGCGTTAAGTATGCCGCCTGAGGAGTACACTCGCAAGGGTGAAACTCAAAAGAATTGACGGGGGCCCGCACAAGCAGTGGAGCATGTGGTTTAATTCGATGCAACGCGAAGGACCTTACCTGGGTTTGACATGTATATGACCGCGGCAGAAATGTCGTTTTCCGC-----AAG--GACATATACACAGGTGCTGCATGGCTGTCGTCAGCTCGTGCCGTGAGGTGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTATCGTTAGTTGCCAGCACTTAGGG-TGGGAACTCTAACGAGACTGCCTGGGTTAACCAG-GAGGAAGGCGAGGATGACGTCAAGTCAGCATGGCCCTTATGCCCAGGG-CGACACACGTGCTACAATGGCCAGTA-CAGAAGGTGGCAAGATCGCGAGATGGAGCAAATC-CTCAAAGCTGGCCCCAGTTCGGATTGTAGTCTGCAACTCGACTACATGAAGTCGGAATTGCTAGTAATGGCGTGTCAGCCATAACGCCGTGAATACGTTCCCGGGC-----------------------------------------------------------------------------------------------'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''This code imports aligned sequneces into a pandas dataframe'''\n",
    "file_name = \"16S\"\n",
    "aligned_sequnces = PyClustalW.Read_alignment_file(file_name)\n",
    "aligned_sequnces['offset'].max()\n",
    "aligned_sequnces\n",
    "aligned_frame = pd.merge(core_data, aligned_sequnces, \n",
    "         how='inner', on=None, left_on=None, right_on=None,\n",
    "         left_index=False, right_index=False, sort=True,\n",
    "         suffixes=('_x', '_y'), copy=True, indicator=False,\n",
    "         validate=None)\n",
    "aligned_frame\n",
    "aligned_sequnces.loc[(aligned_frame['Record id'] == 'XX000001.1')]['Aligned 16S Sequence'].item()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                  _____________ NR_037113.1\n",
      "                                 ________________|\n",
      "                                |                |____________ NR_026215.1\n",
      "                               _|\n",
      "                              | |  ______________________ NR_029356.1\n",
      "                              | |_|\n",
      "                              |   |_____________________________ NR_042365.1\n",
      "                              |\n",
      "                   ___________|                          _ JX131632.1\n",
      "                  |           |                        ,|\n",
      "                  |           |                        ||_ NR_118997.2\n",
      "                  |           |              __________|\n",
      "                  |           |             |          |___ NR_114596.1\n",
      "                  |           |             |          |\n",
      "                 ,|           |_____________|          |___ LC145699.1\n",
      "                 ||                         |\n",
      "                 ||                         |  _____________ NR_044823.1\n",
      "                 ||                         |_|\n",
      "                 ||                           |__________ NR_040789.1\n",
      "                 ||\n",
      "             ____||                       __ NR_026079.2\n",
      "            |    ||______________________|\n",
      "            |    |                       |_ NR_104946.1\n",
      "            |    |\n",
      "            |    |       _______________ NR_104936.1\n",
      "    ________|    |  ____|\n",
      "   |        |    |_|    |____________ NR_026206.1\n",
      "   |        |      |\n",
      "   |        |      |________________ NR_026078.1\n",
      "  _|        |\n",
      " | |        |______________________ NR_044682.2\n",
      " | |\n",
      " | |   , AJ605115.1\n",
      " | |   |\n",
      " | |___| XX000001.1\n",
      " |     |\n",
      " |     |_ NR_024570.1\n",
      " |\n",
      " |___ NR_036886.1\n",
      "_|\n",
      " | _ NR_028894.1\n",
      " ||\n",
      " ||__ NR_041749.1\n",
      " |\n",
      " |_ NR_036794.1\n",
      " |\n",
      " |__ NR_102794.2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Generate Dendro of Bacterial Species from 16S Segments\n",
    "file_name = \"amplicons\"\n",
    "writen_alignment = PyClustalW.create_alignment_file(calculated_data, 'Amplicon', file_name)\n",
    "writen_alignment\n",
    "PyClustalW.ClustalW_alignment(str(file_name + '.faa'))\n",
    "#ClustalW_alignment\n",
    "PyClustalW.Phylo_tree(file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "amplicon_alignments = PyClustalW.Read_alignment_file('amplicons')\n",
    "amplicon_alignments['Aligned 16S Sequence']\n",
    "conserved_regions = PyClustalW.conserved_regions(amplicon_alignments, 'Aligned 16S Sequence')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ending\n"
     ]
    }
   ],
   "source": [
    "conserved_regions\n",
    "\n",
    "runs_documentation = []\n",
    "runs_list = []\n",
    "global_flag = True\n",
    "local_flag = True\n",
    "warning_flag = True\n",
    "global_items = 0\n",
    "local_items = 0\n",
    "length = 0\n",
    "end = len(conserved_regions)\n",
    "              \n",
    "while global_flag is True:\n",
    "    runs_list = []\n",
    "    if global_items + 2 < end:\n",
    "        global_flag = True\n",
    "    else:\n",
    "        global_flag = False\n",
    "        runs_list.append(conserved_regions[global_items + 1])\n",
    "        print('ending')\n",
    "        \n",
    "    local_flag = True\n",
    "    warning_flag = True\n",
    "        \n",
    "    while local_flag is True:\n",
    "        runs_list.append(conserved_regions[global_items])\n",
    "        global_items = global_items + 1\n",
    "\n",
    "        if global_items == end:\n",
    "            local_flag = False \n",
    "        \n",
    "        elif (conserved_regions[global_items-1] == '*'\n",
    "            and conserved_regions[global_items] == '*'\n",
    "            and warning_flag is True):\n",
    "            warning_flag = True\n",
    "        \n",
    "        elif (conserved_regions[global_items-1] == '*'\n",
    "            or conserved_regions[global_items] == '*'\n",
    "            and warning_flag is True):\n",
    "            warning_flag = False\n",
    "        \n",
    "        else:\n",
    "            length = length + (len(runs_list))\n",
    "            runs_documentation.append(str(runs_list))\n",
    "            local_flag = False       \n",
    "\n",
    "#runs_documentation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAACAGGAAGAAGCTTGCTCTTTGCTGACG\n",
      "GGATTAGTGGCGAACGGGTGAGTAACACGT\n",
      "AGGGGAGCTTGCTCCCTGGGTGACGAGCGG\n",
      "ATGGAGGGGGATAACTACTGGAAACGGTAGCTAATACCGCATAACGTCGCAAGACCAAAGAGGGGGACCTTCGGGCCTCTTGCCATCGGATGTGCCCAGATGGGA\n",
      "CTGGGTCTAATACCGGATAGGAGCGTCCACCGCATGGTGGGTGTTGGAAAGATTTATCGGTTTTGGATGGACTCGCGGCCTATCAGCTTGTTGGTGAGGTAATGG\n",
      "GGGGGATAACTACTGGAAACGGTAGCTAATACCGCATAACGTCGCAAGACCAAAGAGGGGGACCTTCGGGCCTCTTGCCATCAGATGTGCCCAGATGGGATTAGC\n"
     ]
    }
   ],
   "source": [
    "ROIs = probe.splice_variable_regions(core_data, '16S Sequence', 1, 7)\n",
    "print(ROIs.loc[(ROIs['Record id'] == 'XX000001.1')]['Variable Regions'].item()[0])\n",
    "print(ROIs.loc[(ROIs['Record id'] == 'NR_037113.1')]['Variable Regions'].item()[0])\n",
    "print(ROIs.loc[(ROIs['Record id'] == 'NR_036886.1')]['Variable Regions'].item()[0])\n",
    "print(ROIs.loc[(ROIs['Record id'] == 'XX000001.1')]['Variable Regions'].item()[1])\n",
    "print(ROIs.loc[(ROIs['Record id'] == 'NR_037113.1')]['Variable Regions'].item()[1])\n",
    "print(ROIs.loc[(ROIs['Record id'] == 'NR_036886.1')]['Variable Regions'].item()[1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PyClustalW.create_alignment_file_segment(ROIs, 'Variable Regions', 'Section_1', 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([(0, 75),\n",
       "  (112, 156),\n",
       "  (270, 471),\n",
       "  (536, 618),\n",
       "  (725, 868),\n",
       "  (928, 1038),\n",
       "  (1096, 1176),\n",
       "  (1233, 1305),\n",
       "  (1358, 1501)],\n",
       " [(75, 111),\n",
       "  (156, 269),\n",
       "  (471, 535),\n",
       "  (618, 724),\n",
       "  (868, 927),\n",
       "  (1038, 1095),\n",
       "  (1176, 1232),\n",
       "  (1305, 1357),\n",
       "  (1501, 1546)])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ROI = aligned_frame.loc[(aligned_frame['Record id'] == 'XX000001.1')]['Aligned 16S Sequence'].item()\n",
    "offset = 0\n",
    "variable_regions = ((0,0), (69-offset,99-offset), (137-offset,242-offset), (433-offset,497-offset), \n",
    "                    (576-offset,682-offset), (822-offset,879-offset), (986-offset,1043-offset),\n",
    "                     (1117-offset,1173-offset),(1243-offset,1294-offset), (1435-offset,1465-offset))\n",
    "count_accu = 0\n",
    "updated_variable_regions = []\n",
    "updated_conserved_regions = []\n",
    "\n",
    "for region in range(0, len(variable_regions)-1):\n",
    "    count = sum(map(lambda x : 1 if '-' in x else 0, ROI[variable_regions[region][1]:variable_regions[region+1][0]]))\n",
    "    \n",
    "    updated_conserved_regions.append(tuple((variable_regions[region][1] + count_accu, \n",
    "                                            variable_regions[region+1][0] + count_accu + count)))\n",
    "    count_accu = count_accu + count\n",
    "    \n",
    "    count = sum(map(lambda x : 1 if '-' in x else 0, \n",
    "                    ROI[variable_regions[region + 1][0] + 1:variable_regions[region + 1][1]]))\n",
    "    \n",
    "    updated_variable_regions.append(tuple((variable_regions[region + 1][0] + count_accu, \n",
    "                                           variable_regions[region + 1][1] + count_accu + count)))\n",
    "    count_accu = count_accu + count + 1\n",
    "\n",
    "'GTGGCGAACGGGTGAGTAA'\n",
    "updated_conserved_regions, updated_variable_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TGTCTGGG-AAACTGCCTGATGGAGGGGGA',\n",
       " 'CAAGCA-CAAAGAGGGGGACCTTAGGGCCTCT---TGC-------CATCGGATGTGCCCAGATGGGATTAGCTAGTAGGTGGGGTAACGGCTCACCTAGGCGACG',\n",
       " 'TTGCT--CATTGACGTTACCC-GCAGAAGAAGCACCGGCTAACTCCGTGCCAGCAGCCGCGGTA',\n",
       " 'CCCCGGGCTCAACCTGGGAACTGCATCTGATACTGGCAA-GCTTGAGTCTCGTAGAGGGGGGTAGAATTCCAGGTG-TAGCGGTGAAATGCGTAGAGATCTGGAGG',\n",
       " 'TGGCTTCCGGAN-NTAACGCGTTAAGTCGACCGCCTGGGGAGTACGGCCGCAAGGTT',\n",
       " 'AGAGATGAG--AATGTGCCTTCGGGA--ACCGTGAGACAGGTGCT------------']"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''This code selects the V1-V9 regions of the aligned sequences relative to E.Coli'''\n",
    "\n",
    "offset = 0\n",
    "E_Coli = amplicon_alignments.loc[(amplicon_alignments['Record id'] == 'NR_024570.1')]['Aligned 16S Sequence'].item()\n",
    "\n",
    "variable_regions = ((69-offset,99-offset), (137-offset,242-offset), (433-offset,497-offset), \n",
    "                    (576-offset,682-offset), (822-offset,879-offset), (986-offset,1043-offset),\n",
    "                     (1117-offset,1173-offset),(1243-offset,1294-offset), (1435-offset,1465-offset))\n",
    "\n",
    "variable_regions_breakdown = [E_Coli[variable_regions[0][0]:variable_regions[0][1]],\n",
    "E_Coli[variable_regions[1][0]:variable_regions[1][1]],\n",
    "E_Coli[variable_regions[2][0]:variable_regions[2][1]],\n",
    "E_Coli[variable_regions[3][0]:variable_regions[3][1]],\n",
    "E_Coli[variable_regions[4][0]:variable_regions[4][1]],\n",
    "E_Coli[variable_regions[5][0]:variable_regions[5][1]]]\n",
    "variable_regions_breakdown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TGGCGGACGGGTGAGTAA',\n",
       " 'TAACTACTGGAAACGGTAGCTAATACCGCATAACGTCG',\n",
       " 'ATCCCTAGCTGGTCTGAGAGGATGACCAGCAACACTGGAACTGAGACACGGTCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGCACAATGGGC-GCAAGCCTGATGCAGCCATGCNGCGTGTATGAAGAAGGCCTTCGGGTTGTAAAGTACTTTCAGCGGGGAGGAAGGGAGTAAAGTTAATACCT',\n",
       " 'AT-ACGGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCACGCAGGCGGTTTGTTAAGTCAGATGTGAAAT',\n",
       " 'AATACCGGTGGCGAAGGCGGCCCCCTGGACGAAGACTGACGCTCA-GGTGCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGTCGACTT--GGAGGTTGTGCCCTTGAGGC--G',\n",
       " 'AAAACTCAAATGAATTGACGGGGGCC-GCACAAGCGGTGGAGCATGTGGTTTAATTCGATGCAACGCGAAGAACCTTACCTGGTCTTGACAT-CCACGGAAGTTTTC']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''This code selects the C1-C9 regions of the aligned sequences relative to E.Coli'''\n",
    "\n",
    "offset = 69-18\n",
    "\n",
    "E_Coli = amplicon_alignments.loc[(amplicon_alignments['Record id'] == 'NR_024570.1')]['Aligned 16S Sequence'].item()\n",
    "\n",
    "variable_regions = ((69-offset,99-offset), (137-offset,242-offset), (433-offset,497-offset), \n",
    "                    (576-offset,682-offset), (822-offset,879-offset), (986-offset,1043-offset),\n",
    "                     (1117-offset,1173-offset),(1243-offset,1294-offset), (1435-offset,1465-offset))\n",
    "\n",
    "conserved_regions_breakdown = [E_Coli[0:variable_regions[0][0]],\n",
    "E_Coli[variable_regions[0][1]:variable_regions[1][0]],\n",
    "E_Coli[variable_regions[1][1]:variable_regions[2][0]],\n",
    "E_Coli[variable_regions[2][1]:variable_regions[3][0]],\n",
    "E_Coli[variable_regions[3][1]:variable_regions[4][0]],\n",
    "E_Coli[variable_regions[4][1]:variable_regions[5][0]]]\n",
    "conserved_regions_breakdown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.23333333333333334\n",
      "0.26666666666666666\n",
      "0.578125\n",
      "0.41509433962264153\n",
      "0.49122807017543857\n",
      "0.12280701754385964\n"
     ]
    }
   ],
   "source": [
    "offset = 69-18\n",
    "E_Coli = conserved_regions\n",
    "\n",
    "variable_regions = ((69-offset,99-offset), (137-offset,242-offset), (433-offset,497-offset), \n",
    "                    (576-offset,682-offset), (822-offset,879-offset), (986-offset,1043-offset),\n",
    "                     (1117-offset,1173-offset),(1243-offset,1294-offset), (1435-offset,1465-offset))\n",
    "\n",
    "unconserved_regions_breakdown = [str(E_Coli[variable_regions[0][0]:variable_regions[0][1]]),\n",
    "str(E_Coli[variable_regions[1][0]:variable_regions[1][1]]),\n",
    "str(E_Coli[variable_regions[2][0]:variable_regions[2][1]]),\n",
    "str(E_Coli[variable_regions[3][0]:variable_regions[3][1]]),\n",
    "str(E_Coli[variable_regions[4][0]:variable_regions[4][1]]),\n",
    "str(E_Coli[variable_regions[5][0]:variable_regions[5][1]])]\n",
    "\n",
    "for i in range(0,len(unconserved_regions_breakdown)):\n",
    "    count = sum(map(lambda x : 1 if '*' in x else 0, unconserved_regions_breakdown[i])) \n",
    "    print(-count/(variable_regions[i][0]-variable_regions[i][1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.8888888888888888\n",
      "0.34210526315789475\n",
      "0.5445026178010471\n",
      "0.46835443037974683\n",
      "0.5285714285714286\n",
      "0.6822429906542056\n"
     ]
    }
   ],
   "source": [
    "offset = 69-18\n",
    "E_Coli = conserved_regions\n",
    "\n",
    "variable_regions = ((69-offset,99-offset), (137-offset,242-offset), (433-offset,497-offset), \n",
    "                    (576-offset,682-offset), (822-offset,879-offset), (986-offset,1043-offset),\n",
    "                     (1117-offset,1173-offset),(1243-offset,1294-offset), (1435-offset,1465-offset))\n",
    "\n",
    "conserved_regions_breakdown = [str(E_Coli[0:variable_regions[0][0]]),\n",
    "str(E_Coli[variable_regions[0][1]:variable_regions[1][0]]),\n",
    "str(E_Coli[variable_regions[1][1]:variable_regions[2][0]]),\n",
    "str(E_Coli[variable_regions[2][1]:variable_regions[3][0]]),\n",
    "str(E_Coli[variable_regions[3][1]:variable_regions[4][0]]),\n",
    "str(E_Coli[variable_regions[4][1]:variable_regions[5][0]])]\n",
    "conserved_regions_breakdown\n",
    "\n",
    "count = sum(map(lambda x : 1 if '*' in x else 0, conserved_regions_breakdown[0]))\n",
    "print(-count/(-variable_regions[0][0]))\n",
    "\n",
    "for i in range(1,len(conserved_regions_breakdown)):\n",
    "    count = sum(map(lambda x : 1 if '*' in x else 0, conserved_regions_breakdown[i])) \n",
    "    print(-count/(variable_regions[i-1][1]-variable_regions[i][0]))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"[' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', '*', '*', '*', '*', ' ', '*', '*', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', '*', ' ', ' ', '*', '*', '*', '*', ' ', ' ', '*', '*', '*', '*', '*', '*', '*', ' ', '*', ' ', ' ', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*']\""
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unconserved_regions_breakdown[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"['*', ' ', ' ', ' ', ' ', '*', '*', ' ', '*', ' ', '*', ' ', ' ', '*', '*', '*', '*', '*', '*', '*', '*', ' ', '*', '*', ' ', ' ', '*', ' ', '*', ' ', ' ', '*', '*', '*', ' ', ' ', '*', '*', ' ', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', ' ', ' ', '*', '*', ' ', ' ', '*', '*', '*', '*', ' ', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', '*', ' ', '*', '*', '*', '*', '*', '*', ' ', '*', '*', ' ', ' ', ' ', '*', ' ', '*', '*', '*', '*', ' ', ' ', ' ', ' ', ' ', '*', '*', ' ', ' ', '*', '*', '*', '*', ' ', ' ', ' ', '*', '*', '*', ' ', '*', ' ', ' ', '*', ' ', '*', ' ', '*', '*', '*', ' ', ' ', ' ', '*', '*', ' ', '*', '*', ' ', '*', ' ', ' ', ' ', '*', '*', ' ', ' ', '*', ' ', ' ', ' ', '*', '*', '*', '*', '*', ' ', ' ', ' ', '*', '*', ' ', '*', ' ', ' ', ' ', ' ', ' ', ' ', ' ', '*', '*', ' ', '*', '*', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', '*', ' ', ' ', ' ', ' ']\""
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conserved_regions_breakdown[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Bacterial.compare_letters(calculated_data.loc[(core_data['Species'] == 'Klebsiella.pneumoniae')]['Amplicon'].item(), calculated_data.loc[(core_data['Species'] == 'Enterococcus.faecalis')]['Amplicon'].item(), 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "R = 1.38\n",
    "T = np.arange(65+295, 90+295, 0.1)\n",
    "H = 10\n",
    "S = 1\n",
    "\n",
    "Stability_Factor = np.exp(-(H-S**T))/(R**T)*10**23\n",
    "\n",
    "row = ['AT-TA', 'TA-AT', 'AT-AT', 'GC-AT', 'CG-AT', 'AT-GC', 'AT-CG', 'GC-GC', 'CG-GC', 'GC-CG']\n",
    "col = ['dTij/dlogNA+', 'Tij 1.0M-Na+ C', 'Tij 1.0M-Na+ K', 'Tij0.0745M-Na+ C', 'Delta Hij', 'Delta Sij']\n",
    "\n",
    "Constants = [(21.00, 81.85, 355.01, 58.23, 8.00, 22.53),\n",
    "(20.11, 86.72, 359.88, 64.10, 8.31, 24.64),\n",
    "(19.78, 89.08, 362.24, 66.77, 8.45, 24.86),\n",
    "(17.76, 99.49, 372.65, 79.51, 9.13, 24.50),\n",
    "(17.10, 103.18, 376.34, 83.94, 9.36, 24.87),\n",
    "(16.87, 104.43, 377.59, 85.45, 9.44, 25.00),\n",
    "(16.21, 107.96, 381.12, 89.72, 9.67, 25.37),\n",
    "(14.18, 118.49, 391.65, 102.50, 10.34, 27.52),\n",
    "(13.20, 124.54, 397.70, 109.69, 10.72, 26.95),\n",
    "(13.20, 124.61, 397.77, 109.76, 10.72, 26.95)]\n",
    "\n",
    "Blake_1999 = pd.DataFrame(Constants, columns=col, index=row)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "TA = str('AAAUUGAAGAGUUUGAUCAUGGCUCAGAUUGAACGCUGGCGGCAGGCCUAACACAUGCAAGUCGAACGGUAACAGGAAGAAGCUUGCUCUUUGCUGACGAGUGGCGGACGGGUGAGUAAUGUCUGGGAAACUGCCUGAUGGAGGGGGAUAACUACUGGAAACGGUAGCUAAUACCGCAUAACGUCGCAAGACCAAAGAGGGGGACCUUCGGGCCUCUUGCCAUCGGAUGUGCCCAGAUGGGAUUAGCUAGUAGGUGGGGUAACGGCUCACCUAGGCGACGAUCCCUAGCUGGUCUGAGAGGAUGACCAGCCACACUGGAACUGAGACACGGUCCAGACUCCUACGGGAGGCAGCAGUGGGGAAUAUUGCACAAUGGGCGCAAGCCUGAUGCAGCCAUGCCGCGUGUAUGAAGAAGGCCUUCGGGUUGUAAAGUACUUUCAGCGGGGAGGAAGGGAGUAAAGUUAAUACCUUUGCUCAUUGACGUUACCCGCAGAAGAAGCACCGGCUAACUCCGUGCCAGCAGCCGCGGUAAUACGGAGGGUGCAAGCGUUAAUCGGAAUUACUGGGCGUAAAGCGCACGCAGGCGGUUUGUUAAGUCAGAUGUGAAAUCCCCGGGCUCAACCUGGGAACUGCAUCUGAUACUGGCAAGCUUGAGUCUCGUAGAGGGGGGUAGAAUUCCAGGUGUAGCGGUGAAAUGCGUAGAGAUCUGGAGGAAUACCGGUGGCGAAGGCGGCCCCCUGGACGAAGACUCACGCUCAGGUGCGAAAGCGUGGGGAGCAAACAGGAUUAGAUACCCUGGUAGUCCACGCCGUAAACGAUGUCGACUUGGAGGUUGUGCCCUUGAGGCGUGGCUUCCGGAGCUAACGCGUUAAGUCGACCGCCUGGGGAGUACGGCCGCAAGGUUAAAACUCAAAUGAAUUGACGGGGGCCCGCACAAGCGGUGGAGCAUGUGGUUUAAUUCGAUGCAACGCGAAGAACCUUACCUGGUCUUGACAUCCACGGAAGUUUUCAGAGAUGAGAAUGUGCCUUCGGGAACCGUGAGACAGGUGCUGCAUGGCUGUCGUCAGCUCGUGUUGUGAAAUGUUGGGUUAAGUCCCGCAACGAGCGCAACCCUUAUCCUUUGUUGCCAGCGGUCCGGCCGGGAACUCAA AGGAGACUGCCAGUGAUAAACUGGAGGAAGGUGGGGAUGACGUCAAGUCAUCAUGGCCCUUACGACCAGGGCUACACACGUGCUACAAUGGCGCAUACAAAGAGAAGCGACCUCGCGAGAGCAAGCGGACCUCAUAAAGUGCGUCGUAGUCCGGAUUGGAGUCUGCAACUCGACUCCAUGAAGUCGGAAUCGCUAGUAAUCGUGGAUCAGAAUGCCACGGUGAAUACGUUCCCGGGCCUUGUACACACCGCCCGUCACACCAUGGGAGUGGGUUGCAAAAGAAGUAGGUAGCUUAACCUUCGGGAGGGCGCUUACCACUUUGUGAUUCAUGACUGGGGUGAAGUCGUAACAAGGUAACCGUAGGGGAACCUGCGGUUGGAUCACCUCCUUA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGTAACAGGAAGAAGCTTGCTCTTTGCTGACGAGTGGCGGACGGGTGAGTAATGTCTGGGAAACTGCCTGATGGAGGGGGATAACTACTGGAAACGGTAGCTAATACCGCATAACGTCGCAAGACCAAAGAGGGGGACCTTCGGGCCTCTTGCCATCGGATGTGCCCAGATGGGATTAGCTAGTAGGTGGGGTAACGGCTCACCTAGGCGACGATCCCTAGCTGGTCTGAGAGGATGACCAGCCACACTGGAACTGAGACACGGTCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGCACAATGGGCGCAAGCCTGATGCAGCCATGCCGCGTGTATGAAGAAGGCCTTCGGGTTGTAAAGTACTTTCAGCGGGGAGGAAGGGAGTAAAGTTAATACCTTTGCTCATTGACGTTACCCGCAGAAGAAGCACCGGCTAACTCCGTGCCAGCAGCCGCGGTAATACGGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCACGCAGGCGGTTTGTTAAGTCAGATGTGAAATCCCCGGGCTCAACCTGGGAACTGCATCTGATACTGGCAAGCTTGAGTCTCGTAGAGGGGGGTAGAATTCCAGGTGTAGCGGTGAAATGCGTAGAGATCTGGAGGAATACCGGTGGCGAAGGCGGCCCCCTGGACGAAGACTCACGCTCAGGTGCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGTCGACTTGGAGGTTGTGCCCTTGAGGCGTGGCTTCCGGAGCTAACGCGTTAAGTCGACCGCCTGGGGAGTACGGCCGCAAGGTTAAAACTCAAATGAATTGACGGGGGCCCGCACAAGCGGTGGAGCATGTGGTTTAATTCGATGCAACGCGAAGAACCTTACCTGGTCTTGACATCCACGGAAGTTTTCAGAGATGAGAATGTGCCTTCGGGAACCGTGAGACAGGTGCTGCATGGCTGTCGTCAGCTCGTGTTGTGAAATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTATCCTTTGTTGCCAGCGGTCCGGCCGGGAACTCAA AGGAGACTGCCAGTGATAAACTGGAGGAAGGTGGGGATGACGTCAAGTCATCATGGCCCTTACGACCAGGGCTACACACGTGCTACAATGGCGCATACAAAGAGAAGCGACCTCGCGAGAGCAAGCGGACCTCATAAAGTGCGTCGTAGTCCGGATTGGAGTCTGCAACTCGACTCCATGAAGTCGGAATCGCTAGTAATCGTGGATCAGAATGCCACGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACACCATGGGAGTGGGTTGCAAAAGAAGTAGGTAGCTTAACCTTCGGGAGGGCGCTTACCACTTTGTGATTCATGACTGGGGTGAAGTCGTAACAAGGTAACCGTAGGGGAACCTGCGGTTGGATCACCTCCTTA'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def changeWord(word):\n",
    "    for letter in word:\n",
    "        if letter == \"U\":\n",
    "            word = word.replace(letter,\"T\")\n",
    "    return word\n",
    "TA = changeWord(TA)\n",
    "TA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0, 68)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region = 0\n",
    "count_accu = 0\n",
    "count = sum(map(lambda x : 1 if '-' in x else 0, ROI[variable_regions[region][1]:variable_regions[region+1][0]]))\n",
    "count_accu = count_accu + count\n",
    "updated_conserved_regions = ()\n",
    "alist = (variable_regions[region][1], variable_regions[region+1][0]-1+count_accu)\n",
    "updated_conserved_regions = updated_conserved_regions + alist\n",
    "#count = sum(map(lambda x : 1 if '-' in x else 0, ROI[variable_regions[region+1][0]+1:variable_regions[region+1][1]]))\n",
    "#updated_variable_regions.append[(variable_regions[region+1][0]+count_accu, variable_regions[region+1][1]+count_accu+count)]\n",
    "#count_accu = count_accu + count\n",
    "updated_conserved_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Record id</th>\n",
       "      <th>Variable Regions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Escherichia.coli</td>\n",
       "      <td>AJ605115.1</td>\n",
       "      <td>(--------C-----AGGAAGCAGCTTGCTG, GTCTGGG-AAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Staphylococcus.epidermidis</td>\n",
       "      <td>JX131632.1</td>\n",
       "      <td>(AGCGAA-----CA-GACGAGGAGCTTGCTC, ACGTGGATAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Staphylococcus.carnosus</td>\n",
       "      <td>LC145699.1</td>\n",
       "      <td>(GCCGAA-----CA-GACGAGGAGCTTGCTC, ACGTGGGTAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Escherichia.coli</td>\n",
       "      <td>NR_024570.1</td>\n",
       "      <td>(AACGGTAAC-----AGGAAGCAGCTTGCTG, GTCTGGG-AAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Streptococcus.parasanguinis</td>\n",
       "      <td>NR_024842.1</td>\n",
       "      <td>(AACGCT-----GA-AGCTTGGTGCTTGCAC, GCGTAGGTAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Chlamydia.trachomatis</td>\n",
       "      <td>NR_025888.1</td>\n",
       "      <td>(AACGGA--------------GCAATTGTTT, GCATAGATAATTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pseudomonas.aeruginosa</td>\n",
       "      <td>NR_026078.1</td>\n",
       "      <td>(AGCTTATGA-----AGG---GAGCTTGCC-, GCCTAGG-AATCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Neisseria.gonorrhoeae</td>\n",
       "      <td>NR_026079.2</td>\n",
       "      <td>(GACGGCAGC-----ACAGGGAAGCTTGCTT, ATATCGG-AACGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Acinetobacter.baumannii</td>\n",
       "      <td>NR_026206.1</td>\n",
       "      <td>(AGCGGGGGA-----AGG---TAGCTTGCT-, GCTTAGG-AATCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Corynebacterium.amycolatum</td>\n",
       "      <td>NR_026215.1</td>\n",
       "      <td>(AACGGTAA-------GGCTCCAGCTTGCTG, ACGTGGGTGACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Streptococcus.dysgalactiae</td>\n",
       "      <td>NR_027517.1</td>\n",
       "      <td>(AACGCT-----GA-GGACTGGTGCTTGCAC, GCGTAGGTAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Streptococcus.pyogenes</td>\n",
       "      <td>NR_028598.1</td>\n",
       "      <td>(AACGCT-----GG-CGGC--GTGCCTAATA, GCGTAGGTAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Streptococcus.mitis</td>\n",
       "      <td>NR_028664.1</td>\n",
       "      <td>(AACGCT-----GA-AGGA-GGAGCTTGCTT, GCGTAGGTAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Streptococcus.pneumoniae</td>\n",
       "      <td>NR_028665.1</td>\n",
       "      <td>(AACGCT-----GA-AGGA-GGAGCTTGCTT, GCGTAGGTAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Citrobacter.freundii</td>\n",
       "      <td>NR_028894.1</td>\n",
       "      <td>(AACGGTAGC-----ACAGAGGAGCTTGCTC, GTCTGGG-AAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Flavonifractor.plautii</td>\n",
       "      <td>NR_029356.1</td>\n",
       "      <td>(AACGGGGTGCTCATGACGGAGGATTCGTCC, GCGTGAGGAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Klebsiella.pneumoniae</td>\n",
       "      <td>NR_036794.1</td>\n",
       "      <td>(AGCGGTAGC-----ACAGAG-AGCTTGCTC, GTCTGGG-AAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Serratia.marcescens</td>\n",
       "      <td>NR_036886.1</td>\n",
       "      <td>(AGCGGTAGC-----ACAGGGGAGCTTGCTC, GTCTGGG-AAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Micrococcus.luteus</td>\n",
       "      <td>NR_037113.1</td>\n",
       "      <td>(AACGATGA-------AGC-CCAGCTTGCTG, ACGTGAGTAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Enterococcus.faecalis</td>\n",
       "      <td>NR_040789.1</td>\n",
       "      <td>(AACGCTTCTTTCC-TCCCGAGTGCTTGCAC, ACGTGGGTAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Streptococcus.agalactiae</td>\n",
       "      <td>NR_040821.1</td>\n",
       "      <td>(AACGCT-----GA-GGTTTGGTGTTTACAC, GCGTAGGTAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Agrobacterium.tumefaciens</td>\n",
       "      <td>NR_041396.1</td>\n",
       "      <td>(AACGC------------------CCCGC--, GCGTGGG-AACAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Klebsiella.oxytoca</td>\n",
       "      <td>NR_041749.1</td>\n",
       "      <td>(AACGGTAGC-----ACAGAG-AGCTTGCTC, GTCTGGG-AAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Fusobacterium.necrophorum</td>\n",
       "      <td>NR_042365.1</td>\n",
       "      <td>(ACTCGAGTCTTC--------GGACTTG---, GCGTAAAGAACTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Haemophilus.influenzae</td>\n",
       "      <td>NR_044682.2</td>\n",
       "      <td>(AACGGTAGC-----AGGAGAAAGCTTGCTT, GCTTGGG-AATCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Listeria.monocytogenes</td>\n",
       "      <td>NR_044823.1</td>\n",
       "      <td>(AACGAA-----CG-GAGGAAGAGCTTGCTC, ACGTGGGCAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Bifidobacterium.adolescentis</td>\n",
       "      <td>NR_074802.2</td>\n",
       "      <td>(AACGGGATC------GGCTGGAGCTTGCTC, GCGTGACCGACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Enterobacter.cloacae</td>\n",
       "      <td>NR_102794.2</td>\n",
       "      <td>(AACGGTAGC-----ACAGAG-AGCTTGCTC, GTCTGGG-AAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Moraxella.osloensis</td>\n",
       "      <td>NR_104936.1</td>\n",
       "      <td>(AACGATGAC-----TCTC--TAGCTTGCT-, ATTTAGG-AATCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Neisseria.meningitidis</td>\n",
       "      <td>NR_104946.1</td>\n",
       "      <td>(-ACGGCAGC-----ACAGAGAAGCTTGCTT, ATATCGG-AACGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Staphylococcus.pettenkoferi</td>\n",
       "      <td>NR_114596.1</td>\n",
       "      <td>(AGCGAA-----CA-GACGGGGAGCTTGCTC, ACGTGGGTAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Staphylococcus.aureus</td>\n",
       "      <td>NR_118997.2</td>\n",
       "      <td>(AGCGAA-----CG-GACGAGAAGCTTGCTT, ACGTGGATAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Escherichia.coli</td>\n",
       "      <td>XX000001.1</td>\n",
       "      <td>(AACGGTAAC-----AGGAAGAAGCTTGCT-, GTCTGGG-AAACT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Species    Record id  \\\n",
       "0               Escherichia.coli   AJ605115.1   \n",
       "1     Staphylococcus.epidermidis   JX131632.1   \n",
       "2        Staphylococcus.carnosus   LC145699.1   \n",
       "3               Escherichia.coli  NR_024570.1   \n",
       "4    Streptococcus.parasanguinis  NR_024842.1   \n",
       "5          Chlamydia.trachomatis  NR_025888.1   \n",
       "6         Pseudomonas.aeruginosa  NR_026078.1   \n",
       "7          Neisseria.gonorrhoeae  NR_026079.2   \n",
       "8        Acinetobacter.baumannii  NR_026206.1   \n",
       "9     Corynebacterium.amycolatum  NR_026215.1   \n",
       "10    Streptococcus.dysgalactiae  NR_027517.1   \n",
       "11        Streptococcus.pyogenes  NR_028598.1   \n",
       "12           Streptococcus.mitis  NR_028664.1   \n",
       "13      Streptococcus.pneumoniae  NR_028665.1   \n",
       "14          Citrobacter.freundii  NR_028894.1   \n",
       "15        Flavonifractor.plautii  NR_029356.1   \n",
       "16         Klebsiella.pneumoniae  NR_036794.1   \n",
       "17           Serratia.marcescens  NR_036886.1   \n",
       "18            Micrococcus.luteus  NR_037113.1   \n",
       "19         Enterococcus.faecalis  NR_040789.1   \n",
       "20      Streptococcus.agalactiae  NR_040821.1   \n",
       "21     Agrobacterium.tumefaciens  NR_041396.1   \n",
       "22            Klebsiella.oxytoca  NR_041749.1   \n",
       "23     Fusobacterium.necrophorum  NR_042365.1   \n",
       "24        Haemophilus.influenzae  NR_044682.2   \n",
       "25        Listeria.monocytogenes  NR_044823.1   \n",
       "26  Bifidobacterium.adolescentis  NR_074802.2   \n",
       "27          Enterobacter.cloacae  NR_102794.2   \n",
       "28           Moraxella.osloensis  NR_104936.1   \n",
       "29        Neisseria.meningitidis  NR_104946.1   \n",
       "30   Staphylococcus.pettenkoferi  NR_114596.1   \n",
       "31         Staphylococcus.aureus  NR_118997.2   \n",
       "32              Escherichia.coli   XX000001.1   \n",
       "\n",
       "                                     Variable Regions  \n",
       "0   (--------C-----AGGAAGCAGCTTGCTG, GTCTGGG-AAACT...  \n",
       "1   (AGCGAA-----CA-GACGAGGAGCTTGCTC, ACGTGGATAACCT...  \n",
       "2   (GCCGAA-----CA-GACGAGGAGCTTGCTC, ACGTGGGTAACCT...  \n",
       "3   (AACGGTAAC-----AGGAAGCAGCTTGCTG, GTCTGGG-AAACT...  \n",
       "4   (AACGCT-----GA-AGCTTGGTGCTTGCAC, GCGTAGGTAACCT...  \n",
       "5   (AACGGA--------------GCAATTGTTT, GCATAGATAATTT...  \n",
       "6   (AGCTTATGA-----AGG---GAGCTTGCC-, GCCTAGG-AATCT...  \n",
       "7   (GACGGCAGC-----ACAGGGAAGCTTGCTT, ATATCGG-AACGT...  \n",
       "8   (AGCGGGGGA-----AGG---TAGCTTGCT-, GCTTAGG-AATCT...  \n",
       "9   (AACGGTAA-------GGCTCCAGCTTGCTG, ACGTGGGTGACCT...  \n",
       "10  (AACGCT-----GA-GGACTGGTGCTTGCAC, GCGTAGGTAACCT...  \n",
       "11  (AACGCT-----GG-CGGC--GTGCCTAATA, GCGTAGGTAACCT...  \n",
       "12  (AACGCT-----GA-AGGA-GGAGCTTGCTT, GCGTAGGTAACCT...  \n",
       "13  (AACGCT-----GA-AGGA-GGAGCTTGCTT, GCGTAGGTAACCT...  \n",
       "14  (AACGGTAGC-----ACAGAGGAGCTTGCTC, GTCTGGG-AAACT...  \n",
       "15  (AACGGGGTGCTCATGACGGAGGATTCGTCC, GCGTGAGGAACCT...  \n",
       "16  (AGCGGTAGC-----ACAGAG-AGCTTGCTC, GTCTGGG-AAACT...  \n",
       "17  (AGCGGTAGC-----ACAGGGGAGCTTGCTC, GTCTGGG-AAACT...  \n",
       "18  (AACGATGA-------AGC-CCAGCTTGCTG, ACGTGAGTAACCT...  \n",
       "19  (AACGCTTCTTTCC-TCCCGAGTGCTTGCAC, ACGTGGGTAACCT...  \n",
       "20  (AACGCT-----GA-GGTTTGGTGTTTACAC, GCGTAGGTAACCT...  \n",
       "21  (AACGC------------------CCCGC--, GCGTGGG-AACAT...  \n",
       "22  (AACGGTAGC-----ACAGAG-AGCTTGCTC, GTCTGGG-AAACT...  \n",
       "23  (ACTCGAGTCTTC--------GGACTTG---, GCGTAAAGAACTT...  \n",
       "24  (AACGGTAGC-----AGGAGAAAGCTTGCTT, GCTTGGG-AATCT...  \n",
       "25  (AACGAA-----CG-GAGGAAGAGCTTGCTC, ACGTGGGCAACCT...  \n",
       "26  (AACGGGATC------GGCTGGAGCTTGCTC, GCGTGACCGACCT...  \n",
       "27  (AACGGTAGC-----ACAGAG-AGCTTGCTC, GTCTGGG-AAACT...  \n",
       "28  (AACGATGAC-----TCTC--TAGCTTGCT-, ATTTAGG-AATCT...  \n",
       "29  (-ACGGCAGC-----ACAGAGAAGCTTGCTT, ATATCGG-AACGT...  \n",
       "30  (AGCGAA-----CA-GACGGGGAGCTTGCTC, ACGTGGGTAACCT...  \n",
       "31  (AGCGAA-----CG-GACGAGAAGCTTGCTT, ACGTGGATAACCT...  \n",
       "32  (AACGGTAAC-----AGGAAGAAGCTTGCT-, GTCTGGG-AAACT...  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conserved_region = []\n",
    "variable_region = []\n",
    "pd_frame = aligned_frame\n",
    "column = 'Aligned 16S Sequence'\n",
    "end_reg = 6\n",
    "start_reg = 1\n",
    "\n",
    "offset = 0\n",
    "variable_regions = ((0,0), (69-offset,99-offset), (137-offset,242-offset), (433-offset,497-offset), \n",
    "                    (576-offset,682-offset), (822-offset,879-offset), (986-offset,1043-offset),\n",
    "                     (1117-offset,1173-offset),(1243-offset,1294-offset), (1435-offset,1465-offset))\n",
    "\n",
    "for item in pd_frame['Record id']:\n",
    "    sequence = pd_frame.loc[(pd_frame['Record id'] == item)][column].item()\n",
    "    the_tuple = ()\n",
    "    for region in range(0, end_reg-start_reg+1):\n",
    "        the_tuple = the_tuple + (sequence[variable_regions[region + 1][0]:variable_regions[region + 1][1]], )\n",
    "    variable_region.append([core_data.loc[(core_data['Record id'] == item)]['Species'].item(),\n",
    "                           item,\n",
    "                           the_tuple])\n",
    "columns = ['Species', 'Record id', 'Variable Regions',]\n",
    "returned_pd_frame = pd.DataFrame(variable_region, columns=columns)\n",
    "returned_pd_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "conserved_regions_adjusted= [(0, 75), \n",
    "                             (112, 156),\n",
    "                             (270, 471),\n",
    "                             (536, 618), \n",
    "                             (725, 868), \n",
    "                             (928, 1038), \n",
    "                             (1096, 1176), \n",
    "                             (1233, 1305),\n",
    "                             (1358, 1501)]\n",
    "variable_regions_adjusted = [(75, 111), \n",
    "                             (156, 269), \n",
    "                             (471, 535), \n",
    "                             (618, 724), \n",
    "                             (868, 927), \n",
    "                             (1038, 1095), \n",
    "                             (1176, 1232), \n",
    "                             (1305, 1357), \n",
    "                             (1501, 1546)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "111"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
